File: src/java/com/distributedlife/mahjong/reference/permute/MixedPungPermutator.java
package com.distributedlife.mahjong.reference.permute;

import com.distributedlife.mahjong.reference.data.TileSet;
import com.distributedlife.mahjong.reference.hand.HandCandidate;

import java.util.*;

public class MixedPungPermutator extends Permutator {
    private List<String> tiles;

    public MixedPungPermutator(List<String> tiles) {
        super();
        this.tiles = tiles;
    }

    private List<List<String>> produceListOfMixedPungs() {
        List<List<String>> listOfMixedPungs = new ArrayList<List<String>>();

        for(String tile : tiles) {
            String bamboo = TileSet.createTile(tile, "Bamboo");
            String crack = TileSet.createTile(tile, "Crack");
            String spot = TileSet.createTile(tile, "Spot");

            listOfMixedPungs.add(Arrays.asList(bamboo, crack, spot));
        }

        return listOfMixedPungs;
    }

    @Override
    Set<HandCandidate> permute(Set<HandCandidate> candidates) {
        Set<HandCandidate> newCandidates = new HashSet<HandCandidate>();

        for (List<String> mixedPung : produceListOfMixedPungs()) {
            for(HandCandidate candidate : candidates) {
                HandCandidate variation = candidate.fork();

                for (String tile : mixedPung) {
                    if (variation.canAdd(tile)) {
                        variation.add(tile);
                    }
                }

                if (variation.getRequiredTiles().size() == candidate.getRequiredTiles().size() + 3) {
                    newCandidates.add(variation);
                }
            }
        }

        return newCandidates;
    }
}

File: lib/permute/mixed-pung-permutator.js
/**
 * MixedPungPermutator - Handles pungs from same rank across multiple suits
 */
import { Permutator } from './permutator.js';
import { createTile } from '../data/tile-set.js';

export class MixedPungPermutator extends Permutator {
  constructor(tiles) {
    super();
    this.tiles = tiles;
  }

  permute(candidates) {
    const newCandidates = new Set();
    const mixedPungs = this.produceListOfMixedPungs();

    for (const mixedPung of mixedPungs) {
      for (const candidate of candidates) {
        const variation = candidate.fork();

        for (const tile of mixedPung) {
          if (variation.canAdd(tile)) {
            variation.add(tile);
          }
        }

        if (variation.getRequiredTiles().length === candidate.getRequiredTiles().length + 3) {
          newCandidates.add(variation);
        }
      }
    }

    return newCandidates;
  }

  produceListOfMixedPungs() {
    const listOfMixedPungs = [];

    for (const tile of this.tiles) {
      const bamboo = createTile(tile, 'Bamboo');
      const crack = createTile(tile, 'Crack');
      const spot = createTile(tile, 'Spot');

      listOfMixedPungs.push([bamboo, crack, spot]);
    }

    return listOfMixedPungs;
  }
}
